File: c++/src/lac.cpp
/* Copyright (c) 2020 Baidu, Inc. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. */

#include "lac.h"
#include "lac_util.h"
#include "paddle_api.h"
#include "lac_custom.h"

/* LAC构造函数：初始化、装载模型和词典 */
LAC::LAC(const std::string& model_path, CODE_TYPE type)
    : _codetype(type),
      _lod(std::vector<std::vector<size_t> >(1)),
      _id2label_dict(new std::unordered_map<int64_t, std::string>),
      _q2b_dict(new std::unordered_map<std::string, std::string>),
      _word2id_dict(new std::unordered_map<std::string, int64_t>),
      custom(NULL)
{

    // 装载词典
    std::string word_dict_path = model_path + "/conf/word.dic";
    load_word2id_dict(word_dict_path, *_word2id_dict);
    std::string q2b_dict_path = model_path + "/conf/q2b.dic";
    load_q2b_dict(q2b_dict_path, *_q2b_dict);
    std::string label_dict_path = model_path + "/conf/tag.dic";
    load_id2label_dict(label_dict_path, *_id2label_dict);

    // 使用AnalysisConfig装载模型，会进一步优化模型
    this->_place = paddle::PaddlePlace::kCPU;
    paddle::AnalysisConfig config;
    // config.SwitchIrOptim(false);       // 关闭优化
    config.EnableMKLDNN();
    config.SetModel(model_path + "/model");
    config.DisableGpu();
    config.SetCpuMathLibraryNumThreads(1);
    config.SwitchUseFeedFetchOps(false);
    this->_predictor = paddle::CreatePaddlePredictor<paddle::AnalysisConfig>(config);

    // 初始化输入输出变量
    auto input_names = this->_predictor->GetInputNames();
    this->_input_tensor = this->_predictor->GetInputTensor(input_names[0]);
    auto output_names = this->_predictor->GetOutputNames();
    this->_output_tensor = this->_predictor->GetOutputTensor(output_names[0]);
    this->_oov_id = this->_word2id_dict->size() - 1;
    auto word_iter = this->_word2id_dict->find("OOV");
    if (word_iter != this->_word2id_dict->end())
    {
        this->_oov_id = word_iter->second;
    }
}

/* 拷贝构造函数，用于多线程重载 */
LAC::LAC(LAC &lac)
    : _codetype(lac._codetype),
      _lod(std::vector<std::vector<size_t> >(1)),
      _id2label_dict(lac._id2label_dict),
      _q2b_dict(lac._q2b_dict),
      _word2id_dict(lac._word2id_dict),
      _oov_id(lac._oov_id),
      _place(lac._place),
      _predictor(lac._predictor->Clone()),
      custom(lac.custom)
{
    auto input_names = this->_predictor->GetInputNames();
    this->_input_tensor = this->_predictor->GetInputTensor(input_names[0]);
    auto output_names = this->_predictor->GetOutputNames();
    this->_output_tensor = this->_predictor->GetOutputTensor(output_names[0]);
}

/* 装载用户词典 */
int LAC::load_customization(const std::string& filename){
    /* 多线程热加载时容易出问题，多个线程共享custom
    if (custom){
        return custom->load_dict(filename);
    }
    */
    custom = std::make_shared<Customization>(filename);
    return 0;
}

/* 将字符串输入转为Tensor */
int LAC::feed_data(const std::vector<std::string> &querys)
{
    this->_seq_words_batch.clear();
    this->_lod[0].clear();

    this->_lod[0].push_back(0);
    int shape = 0;
    for (size_t i = 0; i < querys.size(); ++i)
    {
        split_words(querys[i], this->_codetype, this->_seq_words);
        this->_seq_words_batch.push_back(this->_seq_words);
        shape += this->_seq_words.size();
        this->_lod[0].push_back(shape);
    }
    this->_input_tensor->SetLoD(this->_lod);
    this->_input_tensor->Reshape({shape, 1});

    int64_t *input_d = this->_input_tensor->mutable_data<int64_t>(this->_place);
    int index = 0;
    for (size_t i = 0; i < this->_seq_words_batch.size(); ++i)
    {
        for (size_t j = 0; j < this->_seq_words_batch[i].size(); ++j)
        {
            // normalization
            std::string word = this->_seq_words_batch[i][j];
            auto q2b_iter = this->_q2b_dict->find(word);
            if (q2b_iter != this->_q2b_dict->end())
            {
                word = q2b_iter->second;
            }

            // get word_id
            int64_t word_id = this->_oov_id;
            auto word_iter = this->_word2id_dict->find(word);
            if (word_iter != this->_word2id_dict->end())
            {
                word_id = word_iter->second;
            }
            input_d[index++] = word_id;
        }
    }
    return 0;
}

/* 对输出的标签进行解码转换为模型输出格式 */
int LAC::parse_targets(
    const std::vector<std::string> &tags,
    const std::vector<std::string> &words,
    std::vector<OutputItem> &result)
{
    result.clear();
    for (size_t i = 0; i < tags.size(); ++i)
    {
        // 若新词，则push_back一个新词，否则append到上一个词中
        if (result.empty() || tags[i].rfind("B") == tags[i].length() - 1 || tags[i].rfind("S") == tags[i].length() - 1)
        {
            OutputItem output_item;
            output_item.word = words[i];
            output_item.tag = tags[i].substr(0, tags[i].length() - 2);
            result.push_back(output_item);
        }
        else
        {
            result[result.size() - 1].word += words[i];
        }
    }
    return 0;
}

std::vector<OutputItem> LAC::run(const std::string &query)
{
    std::vector<std::string> query_vector = std::vector<std::string>({query});
    auto result = run(query_vector);
    return result[0];
}

std::vector<std::vector<OutputItem>> LAC::run(const std::vector<std::string> &querys)
{

    this->feed_data(querys);
    this->_predictor->ZeroCopyRun();

    // 对模型输出进行解码
    int output_size = 0;
    int64_t *output_d = this->_output_tensor->data<int64_t>(&(this->_place), &output_size);
    this->_labels.clear();
    this->_results_batch.clear();
    for (size_t i = 0; i < this->_lod[0].size() - 1; ++i)
    {
        for (size_t j = 0; j < _lod[0][i + 1] - _lod[0][i]; ++j)
        {

            int64_t cur_label_id = output_d[_lod[0][i] + j];
            auto it = this->_id2label_dict->find(cur_label_id);
            this->_labels.push_back(it->second);
        }

        // 装载了用户干预词典，先进行干预处理
        if (custom){
            custom->parse_customization(this->_seq_words_batch[i], this->_labels);
        }

        parse_targets(this->_labels, this->_seq_words_batch[i], this->_results);
        this->_labels.clear();
        _results_batch.push_back(this->_results);
    }

    return this->_results_batch;
}

File: python/LAC/lac.py
# -*- coding: UTF-8 -*-
################################################################################
#
#   Copyright (c) 2020  Baidu, Inc. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License"
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#################################################################################

"""
本文件定义了LAC类，实现其调用分词，词性标注，训练模型的接口。
"""

import os
import shutil
import logging

import numpy as np
import paddle.fluid as fluid
from paddle.fluid.core import PaddleTensor
from paddle.fluid.core import AnalysisConfig
from paddle.fluid.core import create_paddle_predictor

from . import reader
from . import utils
from . import nets

from ._compat import *
from .custom import Customization
from .models import Model, SegModel, LacModel, RankModel

def _get_abs_path(path): return os.path.normpath(
    os.path.join(os.getcwd(), os.path.dirname(__file__), path))

DEFAULT_LAC = _get_abs_path('lac_model')
DEFAULT_SEG = _get_abs_path('seg_model')
DEFAULT_RANK = _get_abs_path('rank_model')

PATH_DICT = {
                "lac": DEFAULT_LAC,
                "seg": DEFAULT_SEG,
                "rank":DEFAULT_RANK
             }


class LAC(object):
    """Docstring for LAC"""
    def __init__(self, model_path=None, mode='lac', use_cuda=False):
        super(LAC, self).__init__()
        utils.check_cuda(use_cuda)

        model_path = model_path if model_path else PATH_DICT[mode]

        if mode == 'seg':
            model = SegModel(model_path, mode, use_cuda)
        elif mode == 'lac':
            model = LacModel(model_path, mode, use_cuda)
        elif mode == 'rank':
            model = RankModel(model_path, mode, use_cuda)

        self.model = model

    def run(self, texts):
        """执行模型预测过程
        Args:
            texts: 模型输入的文本，一个Unicode编码的字符串或者
                   由Unicode编码字符串组成的List
        Returns:
            if mode=='seg',  返回分词结果
            if mode=='lac',  返回分词,词性结果
            if mode=='rank', 返回分词,词性,词语重要性结果
        """
        return self.model.run(texts)
    
    def train(self, model_save_dir, train_data, test_data=None, iter_num=10, thread_num=10):
        """执行模型增量训练
        Args:
            model_save_dir: 训练结束后模型保存的路径
            train_data: 训练数据路径
            test_data: 测试数据路径，若为None则不进行测试
            iter_num: 训练数据的迭代次数
            thread_num: 执行训练的线程数
        """
        self.model.train(model_save_dir, train_data, test_data, iter_num, thread_num)
    
    def load_customization(self, customization_file, sep=None):
        """装载用户词典

        Args:
            texts: 用户词典路径
            sep: 表示词典中，短语片段的分隔符，默认为空格' '或制表符'\t'
        """
        self.model.custom = Customization()
        self.model.custom.load_customization(customization_file, sep)
    
    def add_word(self, word, sep=None):
        """添加单词，格式与用户词典一致
        Args:
            texts: 用户定义词典，如："春天"、"花 开"、"春天/SEASON"、"花/n 开/v"、
            sep: 表示词典中，短语片段的分隔符，默认为空格' '或制表符'\t'
        """
        if self.model.custom is None:
            self.model.custom = Customization()
        self.model.custom.add_word(word, sep)

if __name__ == "__main__":
    print('######### mode = lac ##############')
    lac = LAC('lac_model')

    test_data = [u'百度是一家高科技公司', u'LAC是一个优秀的分词工具', '']

    print('######### run:list ##############')
    result = lac.run(test_data)
    for res in result:
        print(' '.join(res))

    print('######### run:str ##############')
    result = lac.run(test_data[0])
    print(' '.join(result))

    print('######### run:tag ##############')
    result = lac.run(test_data)
    for i, (sent, tags) in enumerate(result):
        result_list = ['(%s, %s)' % (ch, tag) for ch, tag in zip(sent, tags)]
        print(''.join(result_list))

    print('######### mode = rank ##############')
    lac = LAC('lac_model', mode='rank')

    print('######### run:list ##############')
    result = lac.run(test_data)
    for res in result:
        print(' '.join(res))

    print('######### run:str ##############')
    result = lac.run(test_data[0])
    print(' '.join(result))

    print('######### run:tag ##############')
    result = lac.run(test_data)
    for i, (sent, tags, word_rank) in enumerate(result):
        result_list = ['(%s, %s, %s)' % (ch, tag, rank) for ch, tag, rank in zip(sent, tags, word_rank)]
        print(''.join(result_list))


    # 重训模型
    lac.train(model_save_dir='models_test',
              train_data='./data/train.tsv', test_data='./data/test.tsv')

    print('######### run:list ##############')
    result = lac.run(test_data)
    for res in result:
        print(' '.join(res))
    print('######### run:str ##############')
    result = lac.run(test_data[0])
    print(' '.join(result))
    print('######### run:tag ##############')
    result = lac.run(test_data, mode='lac')
    for i, (sent, tags) in enumerate(result):
        result_list = ['(%s, %s)' % (ch, tag) for ch, tag in zip(sent, tags)]
        print(''.join(result_list))
